### PR TITLE
Create git storage and talk to it via HTTP & gRPC

### PR DESCRIPTION
**Pull request by metalmatze**
The storage binary now exposes a http server that can be used to `git pull` & `git push` and other things possible via http with git. It's going to be accessed via the API which does the authentication before proxying any requests to the storage.

For everything else, like creating, updating & deleting repositories and all the other things that are going to be necessary to do on a git repositoriy, we're talking to this storage binary via gRPC.
The gRPC client & server are in the `/storage` package. Every RPC received is going to be passed to the storage implementation (`/storage/storage.go`) that doesn't know anything about gRPC, by composition.

### CODE REVIEW

File: cmd/storage/main.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import (
+	"log"
+	"os"
+
+	"github.com/urfave/cli"
+)
+
+func main() {
+	app := cli.NewApp()
+	app.Name = "gitpoids-storage"
```
Comment (fortytw2) [inline cmd/storage/main.go]: gitpoids

File: cmd/storage/storage.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	if err != nil {
+		return err
+	}
+
+	lis, err := net.Listen("tcp", storageConfig.GRPCAddr)
+	if err != nil {
+		return fmt.Errorf("failed to create grpc listener: %v", err)
+	}
+
+	var gr group.Group
```
Comment (fortytw2) [inline cmd/storage/storage.go]: awesome stuff

File: repository/store_postgres.go [hunk 3]
```diff
@@ -17,21 +17,23 @@ func NewPostgresStore(db *sql.DB) *Postgres {
...
+		if id == "" {
+			return nil, nil, nil, OwnerNotFoundError
+		}
 
-	owner := &Owner{ID: id}
+		owner.ID = id
+	}
 
-	query = `
+	query := `
```
Comment (fortytw2) [inline repository/store_postgres.go]: you already have ID from the above query, so no need to subquery

File: repository/service.go [hunk 4]
```diff
@@ -12,42 +12,58 @@ var (
...
+	Storage interface {
+		Create(owner string, name string) error
 	}
 
 	// Service to interact with repositories.
 	Service interface {
-		ListByOwnerUsername(string) ([]*Repository, []*Stats, *Owner, error)
-		Find(string, string) (*Repository, *Stats, *Owner, error)
-		Create(ownerID string, repository *Repository) (*Repository, error)
+		List(owner *Owner) ([]*Repository, []*Stats, *Owner, error)
```
Comment (fortytw2) [inline repository/service.go]: these functions are probably best left as you used to have them - taking an *Owner can be really unclear about "what fields need to be set for this function to work", and you can't always work with 100% hydrated models - thoughts?

File: resolver/repository.go [hunk 5]
```diff
@@ -156,12 +156,12 @@ func (r *RepositoryResolver) CreateRepository(ctx context.Context, args struct{
...
 	}
 
-	repo, err := r.repositories.Create(ownerID, repo)
+	owner, err := r.users.Find(sessOwner.ID)
 	if err != nil {
 		return nil, err
 	}
 
-	owner, err := r.users.Find(ownerID)
+	repo, err = r.repositories.Create(&repository.Owner{ID: owner.ID, Username: owner.Username}, repo)
```
Comment (fortytw2) [inline resolver/repository.go]: like this wrt the earlier comment

File: storage/server.go [hunk 6]
```diff
@@ -0,0 +1,19 @@
+package storage
+
+import (
+	"golang.org/x/net/context"
```
Comment (fortytw2) [inline storage/server.go]: x/net/context? type alias for go 1.9+ so you can use stdlib context

File: repository/service.go (hunk 4 above)
Comment (metalmatze) [reply repository/service.go]: That's a really good point.
The problem is, that for some `func`s I need both, the id and username. In the first `func` called I need the ID and in the seconds one I need the username. So I thought about combining both. Not really sure though. Maybe I might just require them both as strings too.

File: storage/server.go (hunk 6 above)
Comment (metalmatze) [reply storage/server.go]: Ah wait. I thought I still need to use this package. But the alias works the other way around too? AWESOME :tada:

File: repository/service.go (hunk 4 above)
Comment (fortytw2) [reply repository/service.go]: lately I've been favoring DB interaction functions that take and return "primitive" types, like https://github.com/fortytw2/hydrocarbon/blob/master/db.go#L77 - in the cases where structs are returned almost always the structs are 100% hydrated. 

if you need either/or, I would instead have two functions, I think

File: storage/server.go (hunk 6 above)
Comment (fortytw2) [reply storage/server.go]: https://github.com/golang/net/blob/master/context/go19.go

**Pull request merged**